### ISSUE TITLE
Updating open type properties

### ISSUE THREAD

**Issue by phatcher** [object/Simple.OData.Client#352]
If I try to update an open type property I get an error from Microsoft.OData.Core...

Expanding on one of the TripPin tests example...

```
[Fact]
public async Task UpdatePersonWithDynamicOpenProperty()
{
	var person = await _client
		.For<PersonWithOpenTypeContainer>("Person")
		.WithProperties(x => x.Properties)
		.Set(new
		{
			UserName = "gregorsamsa",
			FirstName = "Gregor",
			LastName = "Samsa",
			Properties = new Dictionary<string, object> { { "OpenTypeField", "Description" } },
		})
		.InsertEntryAsync();

	await _client
		.For<PersonWithOpenTypeContainer>("Person")
		.WithProperties(x => x.Properties)
		.Key("gregorsamsa")
		.Set(new {
			Properties = new Dictionary<string, object> { { "OpenTypeField", "Foo" } },
		 })
		.UpdateEntryAsync();

	person = await _client
		.For<PersonWithOpenTypeContainer>("Person")
		.WithProperties(x => x.Properties)
		.Key("gregorsamsa")
		.FindEntryAsync();

	Assert.Equal("Foo", person.Properties["OpenTypeField"]);
}
```

This raises an exception

```
Microsoft.OData.Core.ODataException
The property 'OpenTypeField' does not exist on type 'Microsoft.OData.SampleService.Models.TripPin.Person'. Make sure to only use property names that are defined by the type.
   at Microsoft.OData.Core.WriterValidationUtils.ValidatePropertyDefined(String propertyName, IEdmStructuredType owningStructuredType, Boolean throwOnMissingProperty)
   at Microsoft.OData.Core.JsonLight.ODataJsonLightPropertySerializer.WriteProperty(ODataProperty property, IEdmStructuredType owningType, Boolean isTopLevel, Boolean allowStreamProperty, DuplicatePropertyNamesChecker duplicatePropertyNamesChecker, ProjectedPropertiesAnnotation projectedProperties)
   at Microsoft.OData.Core.JsonLight.ODataJsonLightPropertySerializer.WriteProperties(IEdmStructuredType owningType, IEnumerable`1 properties, Boolean isComplexValue, DuplicatePropertyNamesChecker duplicatePropertyNamesChecker, ProjectedPropertiesAnnotation projectedProperties)
   at Microsoft.OData.Core.JsonLight.ODataJsonLightWriter.StartEntry(ODataEntry entry)
   at Microsoft.OData.Core.ODataWriterCore.InterceptException(Action action)
   at Microsoft.OData.Core.ODataWriterCore.WriteStartEntryImplementation(ODataEntry entry)
   at Microsoft.OData.Core.TaskUtils.GetTaskForSynchronousOperation(Action synchronousOperation)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Simple.OData.Client.V4.Adapter.RequestWriter.<WriteEntryContentAsync>d__2.MoveNext() in C:\Devel\OpenSource\Simple.OData.Client\Simple.OData.Client.V4.Adapter\RequestWriter.cs:line 50
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Simple.OData.Client.RequestWriterBase.<CreateUpdateRequestAsync>d__7.MoveNext() in C:\Devel\OpenSource\Simple.OData.Client\Simple.OData.Client.Core\Adapter\RequestWriterBase.cs:line 72
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Simple.OData.Client.ODataClient.<ExecuteUpdateEntryAsync>d__143.MoveNext() in C:\Devel\OpenSource\Simple.OData.Client\Simple.OData.Client.Core\ODataClient.Internals.cs:line 46
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Simple.OData.Client.ODataClient.<UpdateEntryAsync>d__98.MoveNext() in C:\Devel\OpenSource\Simple.OData.Client\Simple.OData.Client.Core\ODataClient.Async.cs:line 1027
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Simple.OData.Client.BoundClient`1.<UpdateEntryAsync>d__19.MoveNext() in C:\Devel\OpenSource\Simple.OData.Client\Simple.OData.Client.Core\Fluent\BoundClient.Async.cs:line 144
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Simple.OData.Client.Tests.TripPinTests.<UpdatePersonWithDynamicOpenProperty>d__27.MoveNext() in C:\Devel\OpenSource\Simple.OData.Client\Simple.OData.Client.IntegrationTests\TripPinTests.cs:line 439
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.TestInvoker`1.<>c__DisplayClass46_1.<<InvokeTestMethodAsync>b__1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.ExecutionTimer.<AggregateAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Xunit.Sdk.ExceptionAggregator.<RunAsync>d__9.MoveNext() in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.core\Sdk\ExceptionAggregator.cs:line 90
```

**Comment by object** [comment]
I managed to reproduce the issue. It was caused by using PATCH method when updating entities with open types. Looks like I should use PUT. Testing the fix...

**Comment by object** [comment]
Fixed, released in 4.24.2.

**Comment by phatcher** [comment]
I get an error with my test...

```
2016/12/27 00:01:26:151 [DEBUG] ApiClient - Template
2016/12/27 00:01:26:350 [DEBUG] ApiClient - GET request: https://localhost:44302/api/odata/admin/$metadata
2016/12/27 00:01:27:106 [DEBUG] ApiClient - Request completed: OK
2016/12/27 00:01:28:007 [DEBUG] ApiClient - POST request: https://localhost:44302/api/odata/admin/Template
2016/12/27 00:01:29:324 [DEBUG] ApiClient - Request completed: Created
2016/12/27 00:01:29:700 [DEBUG] ApiClient - Template: Id=147
2016/12/27 00:01:29:772 [DEBUG] ApiClient - PUT request: https://localhost:44302/api/odata/admin/Template(147)
2016/12/27 00:01:29:849 [DEBUG] ApiClient - Request completed: PreconditionFailed
2016/12/27 00:01:29:864 [ERROR] ApiClient - Simple.OData.Client.WebRequestException: Precondition Failed
```

Controller is fairly straight forward....

```
    public class TemplateController : AuditedODataV4Controller<Template>, IODataV4Controller<Template, int>
    {
...

        [HttpPut]
        [EnableQuery(AllowedQueryOptions = AllowedQueryOptions.All)]
        public async Task<IHttpActionResult> Put([FromODataUri] int key, Template entity)
        {
            return await UpdateAction(key, entity);
        }

...
    }
```
Internals of the base class just use EF to persist the entity, but I thinks its failing way before then; I've tried changing the verb to Post with no effect; I'll try and chase down what's failing but if you have any quick ideas I'd be grateful.

Also looking at https://github.com/OData/WebApi/issues/186, I think Patch should be supported for dynamic types.

**Comment by object** [comment]
Hmm, perhaps I was too quick replacing PATCH with PUT. I will investigate it further.

**Comment by object** [comment]
Try now (4.24.3), now it should be using PATCH.

**Comment by phatcher** [comment]
@object That fixed the dynamic properties update, however I now have an issue with updating (non-dynamic) classes in an inheritance hierarchy 

```
2016/12/27 11:25:52:702 [DEBUG] ApiClient - Company
2016/12/27 11:25:52:976 [DEBUG] ApiClient - GET request: https://localhost:44302/api/odata/sensors/$metadata
2016/12/27 11:25:53:678 [DEBUG] ApiClient - Request completed: OK
2016/12/27 11:25:54:730 [DEBUG] ApiClient - POST request: https://localhost:44302/api/odata/sensors/Party
2016/12/27 11:25:54:811 [DEBUG] ApiClient - Request completed: Created
2016/12/27 11:26:29:824 [DEBUG] ApiClient - Company: Id=56565
2016/12/27 11:27:21:546 [ERROR] ApiClient - Microsoft.OData.Core.ODataException: A primitive value was specified; however, a value of the non-primitive type '' was expected.
```

This is happening in the SaveAsync; library code hasn't changed so I presume this is a regression somehow - I can let you have access to an endpoint if you email me (paulh@atrios.co.uk)

**Comment by object** [comment]
Fixed in version 4.25.2.
